posts/inside-rust/2021-09-01-Splitting-const-generics.md

```diff
@@ -0,0 +1,129 @@
+---
+layout: post
+title: "Splitting the const generics features"
+author: lcnr
+description: "Splitting the const generics features"
+team: The Const Generics Project Group <https://rust-lang.github.io/project-const-generics/>
+---
+
+After the stabilization of the const generics MVP in version 1.51, the const generics project group has continued to
+work on const generics. Large parts of this work were gated behind the feature gates `const_generics` and `const_evaluatable_checked`. As time went on, the
+`const_generics` feature became fairly useless on its own while the name of
+`const_evaluatable_checked` didn't really capture what this feature was intended to do.
+
+To improve this, we have removed the features `const_generics`, `lazy_normalization_consts`, and `const_evaluatable_checked` starting this nightly release. They have been replaced by `feature(adt_const_params)` and `feature(generic_const_exprs)`.
+
+As there is a lot going on with const generics, here's a quick overview of the new - and preexisting - features and how much still needs to be done for them to get stabilized:
+
+### `feature(adt_const_params)`
+
+On stable, only integers, `char` and `bool` are allowed as the types of const parameters. This feature allows additional types, such as `&'static str` and user defined types.
+```rust
+#![feature(adt_const_params)]
+
+#[derive(PartialEq, Eq)]
+enum ImageFormat {
+    Rgb8,
+    Rgba8,
+    // ...c
+}
+
+struct Image<const FORMAT: ImageFormat> {
+    // ...
+}
+
+impl Image<{ ImageFormat::Rgba }> {
+    fn alpha(&self, pixel: PixelLocation) -> u8 {
+        // ...
+    }
+}
+```
+Note that even with this feature, generic const parameter types, such as `struct Foo<T, const N: T> { ... }`, are forbidden.
+While allowing such things is desired, it adds additional complications exceeding our current capacity.
+
+There are still two major blockers for stabilization: 
+
+The first being the [transition to valtrees](https://github.com/rust-lang/rust/pull/83234). Valtrees are a representation of values as trees with integer nodes, simplifiying the way we interact with more complex types.
+
+Additionally, we have to figure out which types we *even want* to allow as const parameter types. This ties into the discussion
+about ["structural match"](https://github.com/rust-lang/rust/issues/74446), which is still ongoing.
+
+While the issues mentioned above are definitely not trivial, I wouldn't be surprised to see this ready for stabilization in a few months.
```

Comment:
This may just be me reading something that isn't there so feel free to disregard, but the change from the plural to singular first person narrator here jumped out to me while reading. Wasn't sure whether this was reflective of there being differing opinions about this within the WG, or if it was just an accidental `I` instead of `we`

Reply:
true, this started out as a more personal blog and I didn't change all the "I" here. Changed it. While this is mostly reflective of my own opinion, I don't think the rest of @rust-lang/project-const-generics has vastly different expectations here